python/cuml/common/base.pyx

```diff
@@ -19,68 +19,123 @@
 # cython: embedsignature = True
 # cython: language_level = 3
 
-
+import cuml
 import cuml.common.handle
 import cuml.common.cuda
 import inspect
 
+from cudf.core import Series, DataFrame
+from cuml.common.array import Array
+from cupy import ndarray as cupyarray
+from numba.cuda import is_cuda_array
+from numpy import ndarray as numpyarray
+
+_input_type_to_str = {
+    numpyarray: 'numpy',
+    cupyarray: 'cupy',
+    Series: 'cudf',
+    DataFrame: 'cudf'
+}
+
 
 class Base:
     """
     Base class for all the ML algos. It handles some of the common operations
     across all algos. Every ML algo class exposed at cython level must inherit
     from this class.
 
+    Typical estimator design using Base requires three main things:
+
+    1. Call the base init method explicitly from inheriting estimators in their
+        init
+    2. Attributes that users will want to acces, and are array-like should
+        use cuml.common.Array and have a preceding underscore `_` before
+        the name the user expects. That way the __getattr__ of Base will
+        convert it automatically to the appropriate output format for the
+        user. For example in DBSCAN the user expects to be able to access
+        model.labels_, so the code actually has an attribute
+        model._labals_ that gets converted at the moment the user accesses
+        labels_ automatically. No need for extra code in inheritting classes
+        as long as they follow that naming convention.
+    3. To appropriately work for outputs mirroring the format of inputs of the
+        user when appropriate, the code in the inheriting estimator must call
+        the method self._check_output_type(input) with input being the data
+        sent by the user that the method's output and model attributes
+        want to mirror. In general this means that the first call in an
+        estimator's predict(self, X), fit(self, X), transform etc. must be to
+        self._check_output_type(X).
+
+    Parameters
+    ----------
+    handle : cuml.Handle
+        If it is None, a new one is created just for this class
```

Comment:
Since this is the heart of where we explain how estimators work, we should expand docs of handle. Maybe:

Specifies the cuml.handle that holds internal CUDA state for computations in this model. Most importantly, this specifies the CUDA stream that will be used for the model's computations, so users can run different models concurrently in different streams by creating handles in several streams. 
If it is None, a new one is created just for this class.